### ISSUE TITLE
Multi-parameters NetworkSet 

### ISSUE THREAD

**Issue by jhillairet** [scikit-rf/scikit-rf#584]
Currently `NetworkSet` does not explicitly support the definition of multiple parameters/dimensions (issue #535). While it is possible to create a NetworkSet and interpolate the results with respect to a parameter (from the [interpolate_from_network](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.networkSet.NetworkSet.interpolate_from_network.html) method), it only concerns a unique parameter/dimension. 

## Purpose and objectives

A desired feature would be to allow creating NetworkSet in which the parameters for which each Network has been obtained (like temperature, voltage or whatever can be changed between the Networks of the NetworkSet) can be set, then to extract or interpolate data for specific parameters.  

prerequisites: 
 * retrocompatibility (do not create a new object NetworkSetX)

below we use the same nomenclature like in [xarray](http://xarray.pydata.org/en/stable/index.html), as xarray is now a defacto  standard for N-dimensional data. So the parameters for which a network can   be stored in a NetworkSet are _dimensions_ and the values of these parameters are the _coordinates_. 

Let be M(=5) Networks which depends on 2 dimensions `a` and `b`:
 * a = 0, b = 10 -> ntwk_0
 * a = 1, b = 10 -> ntwk_1
 * a = 2, b = 10 -> ntwk_2
 * a = 0, b = 20 -> ntwk_3
 * a = 1, b = 20 -> ntwk_4

## Initializations
multi-dimensional parameters NetworkSet _desired_ behaviour pseudo-code
```python
# the parameter definition is made for each Network, prior to make a NetworkSet
all_params = [
        dict(a=0, b=10, c='A'),
        dict(a=1, b=10, c='A'),
        dict(a=2, b=10, c='A'),
        dict(a=0, b=20, c='A'),
        dict(a=1, b=20, c='A')
    ]
# create M dummy networks which all contain a 'params' dictionnary
ntwks_params = [rf.Network(frequency=freq1, 
                                s=np.random.rand(len(freq1),2,2), 
                                name=f'ntwk_{m}',
                                params=params) \
                     for (m, params) in enumerate(all_params) ]

# Create NetworkSet from a list of Network (as usual)
ns_params = rf.NetworkSet(ntwks_params)  
```
the created NetworkSet contains the new additional parameters
```python
ns.dims -> tuple
ns.coords -> dict
```
### from (and to) a (G)MDIF file (ADS)
issue #359 

MDIF (Measurement Data Interchange Format)
MDIF syntax and example file https://awrcorp.com/download/faq/english/docs/Users_Guide/data_file_formats.html#generalized_mdif

```python
rf.NetworkSet.from_mdif(filename)
```

### from MDL format (IC-CAP, keysight)
MDL file syntax description: http://edadownload.software.keysight.com/eedl/iccap/2011_01/pdf/data.pdf

```python
rf.NetworkSet.from_mdl(filename)
```

### from CITI files (keysight)
CITI file description: https://na.support.keysight.com/plts/help/WebHelp/FilePrint/CITIfile_Format.htm

```python
rf.NetworkSet.from_citi(filename)
```

### from a xarray Dataset ?
```python 
rf.NetworkSet.from_dataset(ds)
```
where `ds` structure is TBD.

### Extracting subset of the NetworkSet

```python
# Select all Network(s) having the parameter key 'a' equal to 1 in the NetworkSet ns 
ns.sel('a', 1)

# Selection based on multiple criteria can be made by chaining sel
ns.sel('model', 'A').sel('voltage', 0).sel('temperature', 25)
```

### N-D Interpolations?
```python
ns.interpolate_from_network(dict(a=0.5, b=10), x) -> Network
```

**Comment by jhillairet** [comment]
very open to comments and suggestions of course. I will update it according to the discussion.

**Comment by arsenovic** [comment]
this is a great idea, and i think one of the biggest new features for skrf.  you have outlined it out perfectly.

**Comment by jhillairet** [comment]
@FranzForstmayr, I move the discussion in #585 here:

> At the moment all the dims are stored inside the NetworkSet. Wouldn't it be better to use a 'params' dictionary property in each Network and create the dims in the NetworkSet dynamically? NetworkSet holds it's Networks in a simple list, and is able to query by parameter.

That would simplify the way the Networks are looped in a NetworkSet (in fact, the way it was), but it moves the parameter-definition to the user-side. I mean, users must define the `params` dictionary themselves before creating a NetworkSet (as usual). 

@vik-s, @jrmagers and others who would like to have multi-param NetworkSet, how and when would like to define your parameters? 

PS : Actually, it is already possible to define some parameters in a Network (like `my_ntwk.temperature`) and to filter the Network in a NetworkSet using these parameters as a key. But I'm not sure that it is widely known.

**Comment by FranzForstmayr** [comment]
> PS : Actually, it is already possible to define some parameters in a Network (like `my_ntwk.temperature`) and to filter the Networks in a NetworkSet using these parameters as a key. But I'm not sure that it is widely known.

As a quite regular skrf user, I can confirm that it's not widely used :)

I just think the handling is easier, because i can add the networks to multiple sets, or move the networks to another sets, without the need of handling the params. The Network creation method could be able to parse params from the comments of the sNp file. Or when creating a NetworkSet from a directory (or zip) we could use parts of the filename. The user has to submit a valid regex for parsing them.

**Comment by vik-s** [comment]
@jhillairet - thank you for laying out this so clearly; this should cover multi-dimensional s-parameter datasets without having to re-architect base classes with `xarray`. I'm especially excited for the "MDL" import feature, since that is pretty much what i work with :-)

Defining `params` dictionary property for each network would work for me, and it makes sense to parse out these independent parameters from a s-parameter filename like @FranzForstmayr suggested. Given that `network` already supports `temperature` as a parameter, associating a `params` dict to a network would be logical and consistent perhaps.

**Comment by jhillairet** [comment]
OK. So to be clear, something like this?
```python
params = [
        dict(a=0, b=10, c='A'),
        dict(a=1, b=10, c='A'),
        dict(a=2, b=10, c='A'),
        dict(a=0, b=20, c='A'),
        dict(a=1, b=20, c='A')
    ]
# create M dummy networks which all contain a 'params' dictionnary
ntwks_params = [rf.Network(frequency=self.freq1, 
                                s=np.random.rand(len(self.freq1),2,2), 
                                name=f'ntwk_{m}',
                                params=params) \
                     for (m, params) in enumerate(params) ]

# Create NetworkSet from a list of Network containing a .params dict parameters
ns_params = rf.NetworkSet(ntwks_params)   
```

**Comment by vik-s** [comment]
This looks good to me. It is nice to make the `params` part of the `network` object like you have described, so it can be looked up when dealing with a collection of networks. Even outside of the context of multi-parameter networks, it is helpful to record some key properties associated with the network, such as date, lot, part or product number, etc.

**Comment by FranzForstmayr** [comment]
This looks great!

**Comment by jhillairet** [comment]
The new `.sel(param, value)` method works for simple selection (key, value)

```python
# return None is no match found
ns.sel('XX', 0)  # -> None
ns.sel('a', 500)  # -> None

# Select all Network(s) having the parameter key 'a' equal to 1 in the NetworkSet 
ns.sel('a', 1) # -> ns of len == 2

# Selection based on multiple criteria can be made by chaining sel
ns.sel('model', 'A').sel('voltage', 0).sel('temperature', 25)
```
or for iterable values:

```python
ns.sel('a', range(0,2))  # -> ns of len == 4

ns.sel('a', [0,2])  # ns of len == 4
```

**Comment by vik-s** [comment]
This is nice! Will `sel()` also work with dicts?
```
ns.sel(dict(model='A', voltage=0, temperature=25))
```
It would be nice to access elements of the multi-parameter NetworkSet using the same way it was created.

**Comment by jhillairet** [comment]
> This is nice! Will `sel()` also work with dicts?

Eventually, good idea.

**Comment by FranzForstmayr** [comment]
Is it possible to return an empty NetworkSet instead of None?
Type definition would be easier, and there's no type error raised, if you wan't to iterate over the networks afterward.

**Comment by jhillairet** [comment]
> Is it possible to return an empty NetworkSet instead of None? Type definition would be easier, and there's no type error raised, if you wan't to iterate over the networks afterward.

OK. We need to allow empty NetworkSet then.

**Comment by jhillairet** [comment]
I'm working on interpolating between multiple parameters. Using the example just above, 

* The easy case: the interpolation for set of parameters `{'a': 1.2, 'X': 10}`: interpolates the s-param between the following two Networks: `{'a':1, 'X':10, 'c':'A'}` and `{'a':2, 'X':10, 'c':'A'}`, which turns out to be similar to what is already possible to do with `interpolate_from_network()`

However, there are some tricky cases I would like to discuss:

* what about interpolating for `{'a': 1.2, 'X': 17}` ? 
At the difference of [scipy.interpolate.interpn](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interpn.html#scipy.interpolate.interpn) or [interpolation in xarrays](https://xarray.pydata.org/en/stable/user-guide/interpolation.html#multi-dimensional-interpolation), here for each coordinate (a,X) we have a s-param array of shape (F,N,N)=(101,2,2) instead of a (len(a), len(X))=(3,2) array...

* what do to in case of redundant parameter values?

* which parameter value to choose if not specified ? 
Example: interpolating for {'a': 0.5} could be made either with 'X'=10 or 20...

**Comment by jhillairet** [comment]
Could anyone send me some examples of  MDIF, MDL and CITI for 1, 2 and N-ports ?

**Comment by FranzForstmayr** [comment]
> However, there are some tricky cases I would like to discuss:
> 
> what about interpolating for `{'a': 1.2, 'X': 17}` ?

This is more like extrapolation, I think it's fine to raise an Exception in this case.

> what do to in case of redundant parameter values?

You mean a network is well described with two parameters A and B, and then one would create a parameter C = A. The interpolation will weight each parameter for 1/3, so in fact the parameter has a weight of 2/3.....I'm not sure if this is necessary at the first step.

> which parameter value to choose if not specified ?
We could be strict and require each parameter specified, or use the median value of the unspecified parameters. 
This would mitigate interpolation error, as the mean could be far away from a sample point.

**Comment by jhillairet** [comment]
OK, I propose the following API:
```python
ns.interpolate_from_params(param, x, sub_params, interp_kind)
```
where `sub_params` is a dictionary that can be used to remove ambiguity. 

Example with previous `ns`

```python
ns.interpolate_from_params('a', 0.5)  # --> will fail (ValueError) 
ns.interpolate_from_params('a', 0.5, {'X': 10})  # --> OK
```

**Comment by jhillairet** [comment]
closed with PR #585
